go/client/chat_cli_rendering.go

```diff
@@ -212,55 +237,154 @@ func (v conversationView) show(g *libkb.GlobalContext, showDeviceName bool) erro
 	return nil
 }
 
-type messageFormatter chat1.MessageFromServerOrError
+// TODO what is this doing here?
+func fetchOneMessage(g *libkb.GlobalContext, conversationID chat1.ConversationID, messageID chat1.MessageID) (chat1.MessageFromServerOrError, error) {
+	deflt := chat1.MessageFromServerOrError{}
 
-func (f messageFormatter) authorAndTime(showDeviceName bool) string {
-	m := chat1.MessageFromServerOrError(f)
-	if m.Message == nil {
-		return ""
+	chatClient, err := GetChatLocalClient(g)
+	if err != nil {
+		return deflt, err
 	}
-	t := gregor1.FromTime(m.Message.ServerHeader.Ctime)
-	if showDeviceName {
-		return fmt.Sprintf("%s <%s> %s", m.Message.SenderUsername, m.Message.SenderDeviceName, shortDurationFromNow(t))
+
+	arg := chat1.GetMessagesLocalArg{
+		ConversationID: conversationID,
+		MessageIDs:     []chat1.MessageID{messageID},
+	}
+	res, err := chatClient.GetMessagesLocal(context.TODO(), arg)
+	if err != nil {
+		return deflt, err
 	}
-	return fmt.Sprintf("%s %s", m.Message.SenderUsername, shortDurationFromNow(t))
+	if len(res.Messages) < 0 {
+		return deflt, fmt.Errorf("empty messages list")
+	}
+	return res.Messages[0], nil
+}
+
+const deletedTextCLI = "[deleted]"
+
+// Everything you need to show a message.
+// Takes into account superseding edits and deletions.
+type messageView struct {
+	MessageID chat1.MessageID
+	// Whether to show this message. Show texts, but not edits or deletes.
+	Renderable                  bool
+	AuthorAndTime               string
+	AuthorAndTimeWithDeviceName string
+	Body                        string
+
+	// Used internally for superseeders
+	messageType chat1.MessageType
 }
 
-func (f messageFormatter) body(g *libkb.GlobalContext) (string, error) {
-	m := chat1.MessageFromServerOrError(f)
-	if m.Message != nil {
-		version, err := m.Message.MessagePlaintext.Version()
+// newMessageView extracts from a message the parts for display
+// It may fetch the superseding message. So that for example a TEXT message will show its EDIT text.
+func newMessageView(g *libkb.GlobalContext, conversationID chat1.ConversationID, m chat1.MessageFromServerOrError) (messageView, error) {
+	mv := messageView{}
+
+	if m.Message == nil {
+		if m.UnboxingError != nil {
+			return mv, fmt.Errorf(fmt.Sprintf("<%s>", *m.UnboxingError))
+		}
+		return mv, fmt.Errorf("unexpected data")
+	}
+
+	mv.MessageID = m.Message.ServerHeader.MessageID
+
+	// Check what message superseeds this one.
+	var mvsup *messageView
+	supersededBy := m.Message.ServerHeader.SupersededBy
+	if supersededBy != 0 {
+		msup, err := fetchOneMessage(g, conversationID, supersededBy)
```

Comment:
I'm a bit concerned about doing fetching here in the middle of the rendering logic. It'd be nice if we can move it to `chat_cli_utils.go` but if there's not a clean way to do it I'm OK with this for now as well.